compiler/src/iree/compiler/Codegen/ExternalInterfaces/CPUEncodingExternalModels.cpp

```diff
@@ -62,6 +63,42 @@ namespace {
 // Utilities.
 //===----------------------------------------------------------------------===//
 
+static FailureOr<IREE::Codegen::ScalableTileFlags>
+getScalableTileFlags(linalg::ContractionDimensions cDims,
+                     IREE::Encoding::EncodingAttr encoding,
+                     DictionaryAttr config) {
+  // TODO(egebeysel): I think this isScalable*Enabled flag should be temporary
+  // and the temporary SME flag should probably come next to it.
+  if (!isAArch64(config) || !isScalableVectorizationEnabled())
+    return failure();
+  std::optional<unsigned> mDim =
+      cDims.m.empty() ? std::nullopt
+                      : encoding.mapDimToOperandIndex(cDims.m[0]);
+  std::optional<unsigned> nDim =
+      cDims.n.empty() ? std::nullopt
+                      : encoding.mapDimToOperandIndex(cDims.n[0]);
+  std::optional<unsigned> kDim = encoding.mapDimToOperandIndex(cDims.k[0]);
+  IREE::Codegen::ScalableTileFlags scalableTiles;
+  // TODO(egebeysel): Choosing between SME
+  // & SVE should also probably reside around here. I guess we could also just
+  // choose SME and then have the fallback option implemented with
+  // 2DScalableTo1D pass and also even from 1D to non-scalable. Currently,
+  // hard-coded to pick SVE since SME with DT is not yet supported.
```

Comment:
I would trim this comment:
```suggestion
  // TODO(egebeysel): Add logic for SME.
```

Basically, the following means that SME is on and should be used instead of SVE:
* `--iree-global-opt-enable-early-materialization=false --iree-llvmcpu-target-cpu-features="+sve,+sme".

We shouldn't require any other logic. With this in mind, since we don't support SME yet, it would be good to bail out if `+sme` is added on the command line.

Reply:
Currently doing this, one thing though - do we need the `+sve` flag for SME or is `+sme` enough by itself? Because ... Apple M4

Reply:
Generally speaking, when naming individual CPU target features, name them all / do not rely in implications between them.  In principle we could propagate implied features, but in practice I don't believe that to be currently, consistently happening.

Reply:
> Generally speaking, when naming individual CPU target features, name them all / do not rely in implications between them. In principle we could propagate implied features, but in practice I don't believe that to be currently, consistently happening.

Oh thanks! So you're saying if we need both of the flags for SME (codegen) to be on, then I should name them both. That makes sense. The next question would be if we have such implication between SME & SVE, i.e. if we need `+sve` for SME to be on. Because AFAIK, M4 only happens to support SME and not SVE :)

Reply:
Assuming that M4 does support both SVE and SME, the answer is that yes you should specify both, in the same way as it is useful to mention `+i8mm,+dotprod` and not just `+i8mm`.  On x86 we have a worse version of this where we mention not only all AVX-512 supported features but also AVX, AVX2, FMA below that.
- I don't believe it to be even possible for a CPU to support SME but not SVE.  SME instructions have register operands that are SVE registers Z0-Z31, used for LHS and RHS tiles, separately from the ZA registers introduced by SME for the accumulator. You could check the armv9 architecture reference manual and I would expect it to say that SME implies SVE.

One way out of that is, instead of passing `--iree-llvmcpu-target-cpu-features=...`, you can pass `--iree-llvmcpu-target-cpu=apple-m4`.  It is up to each CPU architecture to implement the semantics of that flag so on some CPU architectures this will fail to populate CPU features. But on Aarch64, this has been implemented in https://github.com/iree-org/iree/pull/19881 so you should be able to just pass `--iree-llvmcpu-target-cpu=apple-m4`.

On top of enabling CPU features, this enables LLVM scheduling models for the given CPU, so in theory this results in better codegen.  In practice it's up to each LLVM target backend.

Reply:
> I don't believe it to be even possible for a CPU to support SME but not SVE

It is :) 

Z0-Z31 registers from SME _may_ (but don't have to) overlap with Z0-Z31 registers from SVE. And M4 is one example where SME was added while SVE was left out (see Wikipedia for [Apple_M4](https://en.wikipedia.org/wiki/Apple_M4)).

> On top of enabling CPU features, this enables LLVM scheduling models for the given CPU, so in theory this results in better codegen.

Side note, I am told that for big OoO cores, LLVM scheduling models are not really used/relevant.

I appreciate that I am not providing any "official" links here, but most of these things are not so visibly documented 😅 

>  do we need the +sve flag for SME or is +sme enough by itself?

Given the above, on M4 you would only use `+sme`. Or, even better, what @bjacob suggested: 

> But on Aarch64, this has been implemented in https://github.com/iree-org/iree/pull/19881 so you should be able to just pass --iree-llvmcpu-target-cpu=apple-m4.

Reply:
Thanks @banach-space , I stand corrected!